<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Формирование-выборок" data-toc-modified-id="Формирование-выборок-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Формирование выборок</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Исследование-моделей" data-toc-modified-id="Исследование-моделей-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследование моделей</a></span><ul class="toc-item"><li><span><a href="#Алгоритм-DecisionTreeClassifier" data-toc-modified-id="Алгоритм-DecisionTreeClassifier-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Алгоритм DecisionTreeClassifier</a></span></li><li><span><a href="#Алгоритм-RandomForestClassifier" data-toc-modified-id="Алгоритм-RandomForestClassifier-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Алгоритм RandomForestClassifier</a></span></li><li><span><a href="#Алгоритм-LogisticRegression" data-toc-modified-id="Алгоритм-LogisticRegression-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Алгоритм LogisticRegression</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Проверка-модели-на-тестовой-выборке" data-toc-modified-id="Проверка-модели-на-тестовой-выборке-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка модели на тестовой выборке</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Проверка-моделей-на-вменяемость" data-toc-modified-id="Проверка-моделей-на-вменяемость-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Проверка моделей на вменяемость</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Итоги-исследования" data-toc-modified-id="Итоги-исследования-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Итоги исследования</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

<a id="section_0"></a>
# Определение перспективного тарифа для телеком-компании

**Описание проекта**

Компания «Мегалайн» — федеральный оператор сотовой связи. Было выявлено, что многие клиенты пользуются архивными тарифами. «Мегалайну» это не выгодно и имеется потребность в построении системы способной проанализировать поведение клиентов и предложить пользователям новый тариф: «Смарт» или «Ультра».

В нашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы. Предположительно, предобработка данных не понадобится, так как была выполнена ранее.

**Цель исследования**:

Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Данная модель должна характеризоваться  максимально большим значением accuracy. Критерий успешности - доля правильных ответов выше 0.75. Значение accuracy проверяется на тестовой выборке.

Разработка модели выполняется по заказу компании «Мегалайн» и будет использована для построении системы, анализирующей поведение клиентов и рекомендующей один из новых тарифов.

**Ход исследования**

Вначале скачаем датасет и изучим данные. Затем разделите исходные данные на обучающую, валидационную и тестовую выборки. Исследуем качество разных моделей, меняя гиперпараметры. Кратко напишем выводы исследования. Проверим качество моделей на тестовой выборке и на вменяемость.
 
Таким образом, работа будет выполнена в пять этапов:
1.	Обзор данных
2.	Формирование выборок.
3.	Исследование моделей.
4.	Проверка качества моделей на тестовой выборке.
5.	Проверка моделей на вменяемость.

**Используемые библиотеки**

Подключим библиотеки для работы с данными, которые будут использованы в ходе проведения исследования.

In [1]:
import pandas as pd
#import matplotlib.pyplot as plt
#import numpy as np
#import math
#from scipy import stats as st
pd.set_option('display.float_format', '{:,.2f}'.format)
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
import warnings
warnings.filterwarnings('ignore')

In [2]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.dummy import DummyClassifier

<a id="section_1"></a>
## Обзор данных

В нашем распоряжении данные о поведении клиентов, которые уже перешли на новые тарифы: «Смарт» и «Ультра». Сообщено, что предобработка данных уже сделана.

**Описание данных**

Каждый объект в наборе данных — это информация о поведении одного пользователя за месяц. Известно:
- сalls — количество звонков,
- minutes — суммарная длительность звонков в минутах,
- messages — количество sms-сообщений,
- mb_used — израсходованный интернет-трафик в Мб,
- is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

Скачаем датасет. Откроем файл с данными и изучим его.

Сохраним данные в переменную `users_behavior`, предусмотрим два варианта размещения данных: в локальной папке и на сервере.

In [3]:
try:
    users_behavior = pd.read_csv('users_behavior.csv')
except:
    users_behavior = pd.read_csv('https://code.s3.yandex.net/datasets/users_behavior.csv')
    
users_behavior.head()

calls  minutes  messages   mb_used  is_ultra
0  40.00   311.90     83.00 19,915.42         0
1  85.00   516.75     56.00 22,696.96         0
2  77.00   467.66     86.00 21,060.45         0
3 106.00   745.53     81.00  8,437.39         1
4  66.00   418.74      1.00 14,502.75         0

Данные читаемы, проблем с кодировкой, разделителями и тп нет, продолжим изучение.

In [4]:
users_behavior.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


В таблице 3214 объектов по 5 признаков у каждого. Пропусков значений нет. Типы данных соответствуют хранимой информации. Данные соответствуют описанию.

In [5]:
users_behavior.describe()

calls  minutes  messages   mb_used  is_ultra
count 3,214.00 3,214.00  3,214.00  3,214.00  3,214.00
mean     63.04   438.21     38.28 17,207.67      0.31
std      33.24   234.57     36.15  7,570.97      0.46
min       0.00     0.00      0.00      0.00      0.00
25%      40.00   274.58      9.00 12,491.90      0.00
50%      62.00   430.60     30.00 16,943.24      0.00
75%      82.00   571.93     57.00 21,424.70      1.00
max     244.00 1,632.06    224.00 49,745.73      1.00

<a id="describe"></a>
Значений, противоречащих здравому смыслу не выявлено. Распределение значений также выглядит правдоподобным.

### Вывод

В таблице 3214 объектов по 5 признаков у каждого. Ошибок в данных не выявлено. Информация о том, что предобработка данных проведена, подтвердилась.

Количество и качество данных позволяют перейти к следующему этапу исследования.

[В начало раздела 1](#section_1)

<a id="section_2"></a>
## Формирование выборок

Разделим исходные данные на обучающую, валидационную и тестовую выборки в соотношении 3:1:1 соответственно.

Сначала отделим от датасета целевой признак.

In [6]:
features = users_behavior.drop(['is_ultra'], axis=1)
target = users_behavior['is_ultra']

Для деления на выборки используем функцию `train_test_split()`. Разделение проведем в два этапа: сначала выделим данные для обучения модели - 60%, затем оставшиеся 40% в равных частях рапределим в валидационную и тестовую выборки.

In [7]:
features_train, features_test_valid, target_train, target_test_valid = train_test_split(features
                                                                                        , target
                                                                                        , test_size=0.4
                                                                                        , random_state=12345)

In [8]:
features_test, features_valid, target_test, target_valid = train_test_split(features_test_valid
                                                                            , target_test_valid
                                                                            , test_size=0.5
                                                                            , random_state=12345)

Проверим результат разбиения.

In [9]:
features_train.shape
target_train.shape
features_valid.shape
target_valid.shape
features_test.shape
target_test.shape

(1928, 4)

(1928,)

(643, 4)

(643,)

(643, 4)

(643,)

### Вывод

Исходные данные разделены на обучающую (1928 объектов), валидационную (643 объекта) и тестовую (643 объекта) выборки.

Можно приступать к построению и сравнению моделей.

[В начало раздела 2](#section_2)

<a id="section_3"></a>
## Исследование моделей

Исследуем качество разных моделей, меняя гиперпараметры. 

Так как предсказывать мы будем качественный признак, то наша задача состоит в построении моделей классификации, а оценивать успешность её выполнения будем по метрике accuracy. 

Для построения моделей будем использовать 3 алгоритма машинного обучения:
- DecisionTreeClassifier
- RandomForestClassifier
- LogisticRegression

### Алгоритм DecisionTreeClassifier

Исследуем влияние на долю правильных предсказаний (accuracy) следующих гиперпераметров:

- max_depth,
- min_samples_split, 
- min_samples_leaf, 

Добавлять в модель их будем последовательно, каждый раз фиксируя время выполнения ячейки.

Вначале посмотрим на результаты модели с гиперпараметрами по умолчанию.

In [10]:
%%time

# DecisionTreeClassifier

best_model_clf = None
best_result_clf = 0


clf = DecisionTreeClassifier(random_state=12345)
clf.fit(features_train, target_train)
result_train = clf.score(features_train, target_train) # качество модели на обучающей выборке
result_valid = clf.score(features_valid, target_valid) # качество модели на валидационной выборке

if result_valid > best_result_clf:
    best_model_clf = clf # сохраним модель
    best_result_clf = result_valid #  сохраним значение accuracy на валидационных данных

print()
print("Модель DecisionTreeClassifier")
print(f'accuracy_train = {result_train}, accuracy_valid = {result_valid}')


Модель DecisionTreeClassifier
accuracy_train = 1.0, accuracy_valid = 0.7309486780715396
Wall time: 15.6 ms


Начнем с изучения влияния гиперпараметра **max_depth**

In [11]:
%%time

# DecisionTreeClassifier

best_model_clf = None
best_result_clf = 0
best_depth_clf = 0
for depth in range(1, 21): # сделаем цикл для max_depth от 1 до 20
    clf = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    clf.fit(features_train, target_train)
    result_train = clf.score(features_train, target_train) # качество модели на обучающей выборке
    result_valid = clf.score(features_valid, target_valid) # качество модели на валидационной выборке
    print(f'max_depth = {depth}, accuracy_train = {result_train}, accuracy_valid = {result_valid}')
    if result_valid > best_result_clf:
        best_model_clf = clf # сохраним наилучшую модель clf
        best_result_clf = result_valid #  сохраним наилучшее значение accuracy на валидационных данных для clf
        best_depth_clf = depth #  сохраним наилучшее значение max_depth на валидационных данных для clf
print()
print("Наилучшая модель DecisionTreeClassifier")
print(f'max_depth = {best_depth_clf}, accuracy_valid = {best_result_clf}')

max_depth = 1, accuracy_train = 0.7577800829875518, accuracy_valid = 0.7356143079315708
max_depth = 2, accuracy_train = 0.7878630705394191, accuracy_valid = 0.7744945567651633
max_depth = 3, accuracy_train = 0.8075726141078838, accuracy_valid = 0.7791601866251944
max_depth = 4, accuracy_train = 0.8106846473029046, accuracy_valid = 0.7744945567651633
max_depth = 5, accuracy_train = 0.8200207468879668, accuracy_valid = 0.7838258164852255
max_depth = 6, accuracy_train = 0.8376556016597511, accuracy_valid = 0.776049766718507
max_depth = 7, accuracy_train = 0.8558091286307054, accuracy_valid = 0.7993779160186625
max_depth = 8, accuracy_train = 0.8625518672199171, accuracy_valid = 0.7931570762052877
max_depth = 9, accuracy_train = 0.8812240663900415, accuracy_valid = 0.7807153965785381
max_depth = 10, accuracy_train = 0.8890041493775933, accuracy_valid = 0.7884914463452566
max_depth = 11, accuracy_train = 0.9066390041493776, accuracy_valid = 0.7744945567651633
max_depth = 12, accuracy_train 

Изучим влияние гиперпараметра **min_samples_split**

In [12]:
%%time

# DecisionTreeClassifier

best_model_clf = None
best_result_clf = 0
best_depth_clf = 0
best_split_clf = 0
for split in range(2, 41): # сделаем цикл для min_samples_split от 2 до 40
    for depth in range(1, 21):
        clf = DecisionTreeClassifier(random_state=12345, max_depth=depth, min_samples_split=split)
        clf.fit(features_train, target_train)
        result_train = clf.score(features_train, target_train) # качество модели на обучающей выборке
        result_valid = clf.score(features_valid, target_valid) # качество модели на валидационной выборке
        if result_valid > best_result_clf:
            best_model_clf = clf # сохраним наилучшую модель clf
            best_result_clf = result_valid #  сохраним наилучшее значение accuracy на валидационных данных для clf
            best_depth_clf = depth # сохраним наилучшее значение max_depth на валидационных данных для clf
            best_split_clf = split # сохраним наилучшее значение min_samples_split на валидационных данных для clf
    print(f'min_samples_split = {best_split_clf}, max_depth = {best_depth_clf}, accuracy_valid = {best_result_clf}')
print()
print("Наилучшая модель DecisionTreeClassifier")
print(f'min_samples_split = {best_split_clf}, max_depth = {best_depth_clf}, accuracy_valid = {best_result_clf}')

min_samples_split = 2, max_depth = 7, accuracy_valid = 0.7993779160186625
min_samples_split = 2, max_depth = 7, accuracy_valid = 0.7993779160186625
min_samples_split = 2, max_depth = 7, accuracy_valid = 0.7993779160186625
min_samples_split = 2, max_depth = 7, accuracy_valid = 0.7993779160186625
min_samples_split = 2, max_depth = 7, accuracy_valid = 0.7993779160186625
min_samples_split = 2, max_depth = 7, accuracy_valid = 0.7993779160186625
min_samples_split = 2, max_depth = 7, accuracy_valid = 0.7993779160186625
min_samples_split = 2, max_depth = 7, accuracy_valid = 0.7993779160186625
min_samples_split = 2, max_depth = 7, accuracy_valid = 0.7993779160186625
min_samples_split = 2, max_depth = 7, accuracy_valid = 0.7993779160186625
min_samples_split = 2, max_depth = 7, accuracy_valid = 0.7993779160186625
min_samples_split = 2, max_depth = 7, accuracy_valid = 0.7993779160186625
min_samples_split = 2, max_depth = 7, accuracy_valid = 0.7993779160186625
min_samples_split = 2, max_depth = 7, 

Изучим также влияние гиперпараметра **min_samples_leaf**

In [13]:
%%time

# DecisionTreeClassifier

best_model_clf = None
best_result_clf = 0
best_depth_clf = 0
best_split_clf = 0
best_leaf_clf = 0
for leaf in range(1, 21): # сделаем цикл для min_samples_leaf от 1 до 20
    for split in range(2, 41):
        for depth in range(1, 21):
            clf = DecisionTreeClassifier(random_state=12345, max_depth=depth, min_samples_split=split, min_samples_leaf=leaf)
            clf.fit(features_train, target_train)
            result_train = clf.score(features_train, target_train) # качество модели на обучающей выборке
            result_valid = clf.score(features_valid, target_valid) # качество модели на валидационной выборке
            if result_valid > best_result_clf:
                best_model_clf = clf # сохраним наилучшую модель clf
                best_result_clf = result_valid #  сохраним наилучшее значение accuracy на валидационных данных для clf
                best_depth_clf = depth # сохраним наилучшее значение max_depth на валидационных данных для clf
                best_split_clf = split # сохраним наилучшее значение min_samples_split на валидационных данных для clf
                best_leaf_clf = leaf # сохраним наилучшее значение min_samples_leaf на валидационных данных для clf
    print(f'min_samples_leaf = {best_leaf_clf}, min_samples_split = {best_split_clf}, max_depth = {best_depth_clf}, accuracy_valid = {best_result_clf}')
print()
print("Наилучшая модель DecisionTreeClassifier")
print(f'min_samples_leaf = {best_leaf_clf}, min_samples_split = {best_split_clf}, max_depth = {best_depth_clf}, accuracy_valid = {best_result_clf}')

min_samples_leaf = 1, min_samples_split = 39, max_depth = 8, accuracy_valid = 0.8040435458786936
min_samples_leaf = 1, min_samples_split = 39, max_depth = 8, accuracy_valid = 0.8040435458786936
min_samples_leaf = 1, min_samples_split = 39, max_depth = 8, accuracy_valid = 0.8040435458786936
min_samples_leaf = 1, min_samples_split = 39, max_depth = 8, accuracy_valid = 0.8040435458786936
min_samples_leaf = 1, min_samples_split = 39, max_depth = 8, accuracy_valid = 0.8040435458786936
min_samples_leaf = 1, min_samples_split = 39, max_depth = 8, accuracy_valid = 0.8040435458786936
min_samples_leaf = 1, min_samples_split = 39, max_depth = 8, accuracy_valid = 0.8040435458786936
min_samples_leaf = 1, min_samples_split = 39, max_depth = 8, accuracy_valid = 0.8040435458786936
min_samples_leaf = 1, min_samples_split = 39, max_depth = 8, accuracy_valid = 0.8040435458786936
min_samples_leaf = 1, min_samples_split = 39, max_depth = 8, accuracy_valid = 0.8040435458786936
min_samples_leaf = 1, min_samp

**Наилучшая модель DecisionTreeClassifier**

Наилучшая модель DecisionTreeClassifier сохранена в переменной `best_model_clf` и имеет следующие 

гиперпераметры:
- min_samples_leaf = 16, 
- min_samples_split = 36, 
- max_depth = 8, 

долю правильных ответов:
- accuracy_valid = 0.807153965785381

была обучена за время:
- 1min 59s

Причем,

модель с гиперпараметрами по умолчанию:
- accuracy_valid = 0.7309486780715396
- Wall time: 12.5 ms

наилучшая модель с перебором 1 гиперпараметра - max_depth - имела результаты:
- max_depth = 7, accuracy_valid = 0.7993779160186625
- Wall time: 183 ms

а наилучшая модель с перебором 2 гиперпараметров - max_depth и min_samples_split - имела результаты:
- min_samples_split = 39, max_depth = 8, accuracy_valid = 0.8040435458786936
- Wall time: 6.26 s

### Алгоритм RandomForestClassifier

Исследуем влияние на долю правильных предсказаний (accuracy) следующих гиперпераметров:

- n_estimators,
- max_depth,
- min_samples_split, 

Добавлять в модель их будем последовательно, каждый раз фиксируя время выполнения ячейки.

Вначале посмотрим на результаты модели с гиперпараметрами по умолчанию.

In [14]:
%%time

# RandomForestClassifier

best_model_clf_2 = None
best_result_clf_2 = 0


clf_2 = RandomForestClassifier(random_state=12345)
clf_2.fit(features_train, target_train)
result_train = clf_2.score(features_train, target_train) # качество модели на обучающей выборке
result_valid = clf_2.score(features_valid, target_valid) # качество модели на валидационной выборке

if result_valid > best_result_clf_2:
    best_model_clf_2 = clf_2 # сохраним наилучшую модель clf_2
    best_result_clf_2 = result_valid #  сохраним наилучшее значение accuracy на валидационных данных для clf_2

print()
print("Модель RandomForestClassifier")
print(f'accuracy_train = {result_train}, accuracy_valid = {result_valid}')


Модель RandomForestClassifier
accuracy_train = 0.9994813278008299, accuracy_valid = 0.7853810264385692
Wall time: 411 ms


Начнем с изучения влияния гиперпараметра **n_estimators**

In [15]:
%%time

# RandomForestClassifier

best_model_clf_2 = None
best_result_clf_2 = 0
best_est_clf_2 = 0

for est in range(10, 201, 10): # сделаем цикл для n_estimators от 10 до 100
    clf_2 = RandomForestClassifier(random_state=12345, n_estimators=est)
    clf_2.fit(features_train, target_train)
    result_train = clf_2.score(features_train, target_train) # качество модели на обучающей выборке
    result_valid = clf_2.score(features_valid, target_valid) # качество модели на валидационной выборке
    print(f'n_estimators = {est}, accuracy_train = {result_train}, accuracy_valid = {result_valid}')
    if result_valid > best_result_clf_2:
        best_model_clf_2 = clf_2 # сохраним наилучшую модель clf_2
        best_result_clf_2 = result_valid #  сохраним наилучшее значение accuracy на валидационных данных для clf_2
        best_est_clf_2 = est # сохраним наилучшее значение n_estimators на валидационных данных для clf_2

print()
print("Наилучшая модель RandomForestClassifier")
print(f'n_estimators = {best_est_clf_2}, accuracy_valid = {best_result_clf_2}')

n_estimators = 10, accuracy_train = 0.9823651452282157, accuracy_valid = 0.7807153965785381
n_estimators = 20, accuracy_train = 0.9891078838174274, accuracy_valid = 0.7822706065318819
n_estimators = 30, accuracy_train = 0.995850622406639, accuracy_valid = 0.7853810264385692
n_estimators = 40, accuracy_train = 0.9974066390041494, accuracy_valid = 0.7931570762052877
n_estimators = 50, accuracy_train = 0.9979253112033195, accuracy_valid = 0.7931570762052877
n_estimators = 60, accuracy_train = 0.9994813278008299, accuracy_valid = 0.7947122861586314
n_estimators = 70, accuracy_train = 0.9994813278008299, accuracy_valid = 0.7931570762052877
n_estimators = 80, accuracy_train = 0.9994813278008299, accuracy_valid = 0.7962674961119751
n_estimators = 90, accuracy_train = 0.9994813278008299, accuracy_valid = 0.7916018662519441
n_estimators = 100, accuracy_train = 0.9994813278008299, accuracy_valid = 0.7853810264385692
n_estimators = 110, accuracy_train = 0.9994813278008299, accuracy_valid = 0.7900

Изучим влияние гиперпараметра **max_depth**

In [16]:
%%time

# RandomForestClassifier

best_model_clf_2 = None
best_result_clf_2 = 0
best_est_clf_2 = 0
best_depth_clf_2 = 0

for depth in range(1, 21): # сделаем цикл для depth от 1 до 20
    for est in range(10, 101, 10):
        clf_2 = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth)
        clf_2.fit(features_train, target_train)
        result_train = clf_2.score(features_train, target_train) # качество модели на обучающей выборке
        result_valid = clf_2.score(features_valid, target_valid) # качество модели на валидационной выборке
        # print(f'min_samples_split = {split}, max_depth = {depth}, accuracy_train = {result_train}, accuracy_valid = {result_valid}')
        if result_valid > best_result_clf_2:
            best_model_clf_2 = clf_2 # сохраним наилучшую модель clf_2
            best_result_clf_2 = result_valid #  сохраним наилучшее значение accuracy на валидационных данных для clf_2
            best_est_clf_2 = est # сохраним наилучшее значение n_estimators на валидационных данных для clf_2
            best_depth_clf_2 = depth # сохраним наилучшее значение max_depth на валидационных данных для clf_2

    print(f'n_estimators = {best_est_clf_2}, max_depth = {best_depth_clf_2}, accuracy_valid = {best_result_clf_2}')
print()
print("Наилучшая модель RandomForestClassifier")
print(f'n_estimators = {best_est_clf_2}, max_depth = {best_depth_clf_2}, accuracy_valid = {best_result_clf_2}')

n_estimators = 40, max_depth = 1, accuracy_valid = 0.7558320373250389
n_estimators = 10, max_depth = 2, accuracy_valid = 0.776049766718507
n_estimators = 30, max_depth = 3, accuracy_valid = 0.7807153965785381
n_estimators = 10, max_depth = 4, accuracy_valid = 0.7853810264385692
n_estimators = 70, max_depth = 5, accuracy_valid = 0.7931570762052877
n_estimators = 30, max_depth = 6, accuracy_valid = 0.7993779160186625
n_estimators = 30, max_depth = 7, accuracy_valid = 0.8040435458786936
n_estimators = 10, max_depth = 8, accuracy_valid = 0.807153965785381
n_estimators = 10, max_depth = 9, accuracy_valid = 0.8133748055987559
n_estimators = 10, max_depth = 9, accuracy_valid = 0.8133748055987559
n_estimators = 10, max_depth = 9, accuracy_valid = 0.8133748055987559
n_estimators = 10, max_depth = 9, accuracy_valid = 0.8133748055987559
n_estimators = 10, max_depth = 9, accuracy_valid = 0.8133748055987559
n_estimators = 10, max_depth = 9, accuracy_valid = 0.8133748055987559
n_estimators = 10, max

Изучим также влияние гиперпараметра **min_samples_split**

In [17]:
%%time

# RandomForestClassifier

best_model_clf_2 = None
best_result_clf_2 = 0
best_est_clf_2 = 0
best_depth_clf_2 = 0
best_split_clf_2 = 0

for split in range(2, 41): # сделаем цикл для min_samples_split от 2 до 40
    for est in range(10, 101, 10):
        for depth in range(1, 21):
            clf_2 = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth, min_samples_split=split)
            clf_2.fit(features_train, target_train)
            result_train = clf_2.score(features_train, target_train) # качество модели на обучающей выборке
            result_valid = clf_2.score(features_valid, target_valid) # качество модели на валидационной выборке
            # print(f'min_samples_split = {split}, max_depth = {depth}, accuracy_train = {result_train}, accuracy_valid = {result_valid}')
            if result_valid > best_result_clf_2:
                best_model_clf_2 = clf_2 # сохраним наилучшую модель clf_2
                best_result_clf_2 = result_valid #  сохраним наилучшее значение accuracy на валидационных данных для clf_2
                best_est_clf_2 = est # сохраним наилучшее значение n_estimators на валидационных данных для clf_2
                best_depth_clf_2 = depth # сохраним наилучшее значение max_depth на валидационных данных для clf_2
                best_split_clf_2 = split # сохраним наилучшее значение min_samples_split на валидационных данных для clf_2

    print(f'n_estimators = {best_est_clf_2}, min_samples_split = {best_split_clf_2}, max_depth = {best_depth_clf_2}, accuracy_valid = {best_result_clf_2}')
print()
print("Наилучшая модель RandomForestClassifier")
print(f'n_estimators = {best_est_clf_2}, min_samples_split = {best_split_clf_2}, max_depth = {best_depth_clf_2}, accuracy_valid = {best_result_clf_2}')

n_estimators = 10, min_samples_split = 2, max_depth = 9, accuracy_valid = 0.8133748055987559
n_estimators = 10, min_samples_split = 2, max_depth = 9, accuracy_valid = 0.8133748055987559
n_estimators = 10, min_samples_split = 2, max_depth = 9, accuracy_valid = 0.8133748055987559
n_estimators = 10, min_samples_split = 5, max_depth = 9, accuracy_valid = 0.8149300155520995
n_estimators = 10, min_samples_split = 5, max_depth = 9, accuracy_valid = 0.8149300155520995
n_estimators = 10, min_samples_split = 5, max_depth = 9, accuracy_valid = 0.8149300155520995
n_estimators = 10, min_samples_split = 5, max_depth = 9, accuracy_valid = 0.8149300155520995
n_estimators = 10, min_samples_split = 5, max_depth = 9, accuracy_valid = 0.8149300155520995
n_estimators = 10, min_samples_split = 5, max_depth = 9, accuracy_valid = 0.8149300155520995
n_estimators = 10, min_samples_split = 5, max_depth = 9, accuracy_valid = 0.8149300155520995
n_estimators = 10, min_samples_split = 5, max_depth = 9, accuracy_vali

**Наилучшая модель RandomForestClassifier**

Наилучшая модель RandomForestClassifier сохранена в переменной `best_model_clf_2` и имеет следующие 

гиперпераметры:
- n_estimators = 20, 
- min_samples_split = 13, 
- max_depth = 11, 

долю правильных ответов:
- accuracy_valid = 0.8180404354587869

была обучена за время:
- 21min 15s

Причем,

модель с гиперпараметрами по умолчанию:
- accuracy_valid = 0.7853810264385692
- Wall time: 409 ms

наилучшая модель с перебором 1 гиперпараметра - n_estimators - имела результаты:
- n_estimators = 80, accuracy_valid = 0.7962674961119751
- Wall time: 8.26 s

а наилучшая модель с перебором 2 гиперпараметров - max_depth и n_estimators - имела результаты:
- n_estimators = 10, max_depth = 9, accuracy_valid = 0.8133748055987559
- Wall time: 34 s

### Алгоритм LogisticRegression

Исследуем влияние на долю правильных предсказаний (accuracy) следующих гиперпераметров:

- max_iter,

Проверить влияние гиперпераметров penalty и solver не удалось, поскольку не все их значения совместимы.

Вначале посмотрим на результаты модели с гиперпараметрами по умолчанию.

In [18]:
%%time

# LogisticRegression

best_model_clf_3 = None
best_result_clf_3 = 0


clf_3 = LogisticRegression(random_state=12345)
clf_3.fit(features_train, target_train)
result_train = clf_3.score(features_train, target_train) # качество модели на обучающей выборке
result_valid = clf_3.score(features_valid, target_valid) # качество модели на валидационной выборке

if result_valid > best_result_clf_3:
    best_model_clf_3 = clf_3 # сохраним наилучшую модель clf_3
    best_result_clf_3 = result_valid #  сохраним наилучшее значение accuracy на валидационных данных для clf_3

print()
print("Модель LogisticRegression")
print(f'accuracy_train = {result_train}, accuracy_valid = {result_valid}')


Модель LogisticRegression
accuracy_train = 0.7131742738589212, accuracy_valid = 0.6842923794712286
Wall time: 98.3 ms


Изучим влияние гиперпараметра **max_iter**

In [19]:
%%time

# LogisticRegression

best_model_clf_3 = None
best_result_clf_3 = 0
best_iter_clf_3 = 0

for iter_ in range(10, 101, 10): # сделаем цикл для max_iter
    clf_3 = LogisticRegression(random_state=12345, max_iter=iter_)
    clf_3.fit(features_train, target_train)
    result_train = clf_3.score(features_train, target_train) # качество модели на обучающей выборке
    result_valid = clf_3.score(features_valid, target_valid) # качество модели на валидационной выборке
    print(f'max_iter = {iter_}, accuracy_train = {result_train}, accuracy_valid = {result_valid}')
    if result_valid > best_result_clf_3:
        best_model_clf_3 = clf_3 # сохраним наилучшую модель clf_3
        best_result_clf_3 = result_valid #  сохраним наилучшее значение accuracy на валидационных данных для clf_3
        best_iter_clf_3 = iter_ # сохраним наилучшее значение n_estimators на валидационных данных для clf_3

    # print(f'max_iter = {best_iter_clf_3}, accuracy_valid = {best_result_clf_3}')
print()
print("Наилучшая модель LogisticRegression")
print(f'max_iter = {best_iter_clf_3}, accuracy_valid = {best_result_clf_3}')

max_iter = 10, accuracy_train = 0.6924273858921162, accuracy_valid = 0.6842923794712286
max_iter = 20, accuracy_train = 0.7152489626556017, accuracy_valid = 0.6967340590979783
max_iter = 30, accuracy_train = 0.7126556016597511, accuracy_valid = 0.6936236391912908
max_iter = 40, accuracy_train = 0.7131742738589212, accuracy_valid = 0.6858475894245724
max_iter = 50, accuracy_train = 0.7131742738589212, accuracy_valid = 0.6842923794712286
max_iter = 60, accuracy_train = 0.7131742738589212, accuracy_valid = 0.6842923794712286
max_iter = 70, accuracy_train = 0.7131742738589212, accuracy_valid = 0.6842923794712286
max_iter = 80, accuracy_train = 0.7131742738589212, accuracy_valid = 0.6842923794712286
max_iter = 90, accuracy_train = 0.7131742738589212, accuracy_valid = 0.6842923794712286
max_iter = 100, accuracy_train = 0.7131742738589212, accuracy_valid = 0.6842923794712286

Наилучшая модель LogisticRegression
max_iter = 20, accuracy_valid = 0.6967340590979783
Wall time: 191 ms


**Наилучшая модель LogisticRegression**

Наилучшая модель LogisticRegression сохранена в переменной `best_model_clf_3` и имеет следующие 

гиперпераметры:
- max_iter = 20, 

долю правильных ответов:
- accuracy_valid = 0.6967340590979783

была обучена за время:
- 178 ms

Причем,

модель с гиперпараметрами по умолчанию:
- accuracy_valid = 0.6842923794712286
- Wall time: 21 ms

### Вывод

Мы обучили модели с помощью трех алгоритмов, сравним получившиеся результаты.

**Наилучшая модель DecisionTreeClassifier**

Наилучшая модель DecisionTreeClassifier сохранена в переменной `best_model_clf` и имеет следующие 

гиперпераметры:
- min_samples_leaf = 16, 
- min_samples_split = 36, 
- max_depth = 8, 

долю правильных ответов:
- accuracy_valid = 0.807153965785381

была обучена за время:
- 1min 59s

Причем,

модель с гиперпараметрами по умолчанию:
- accuracy_valid = 0.7309486780715396
- Wall time: 12.5 ms

наилучшая модель с перебором 1 гиперпараметра - max_depth - имела результаты:
- max_depth = 7, accuracy_valid = 0.7993779160186625
- Wall time: 183 ms

а наилучшая модель с перебором 2 гиперпараметров - max_depth и min_samples_split - имела результаты:
- min_samples_split = 39, max_depth = 8, accuracy_valid = 0.8040435458786936
- Wall time: 6.26 s

**Наилучшая модель RandomForestClassifier**

Наилучшая модель RandomForestClassifier сохранена в переменной `best_model_clf_2` и имеет следующие 

гиперпераметры:
- n_estimators = 20, 
- min_samples_split = 13, 
- max_depth = 11, 

долю правильных ответов:
- accuracy_valid = 0.8180404354587869

была обучена за время:
- 21min 15s

Причем,

модель с гиперпараметрами по умолчанию:
- accuracy_valid = 0.7853810264385692
- Wall time: 409 ms

наилучшая модель с перебором 1 гиперпараметра - n_estimators - имела результаты:
- n_estimators = 80, accuracy_valid = 0.7962674961119751
- Wall time: 8.26 s

а наилучшая модель с перебором 2 гиперпараметров - max_depth и n_estimators - имела результаты:
- n_estimators = 10, max_depth = 9, accuracy_valid = 0.8133748055987559
- Wall time: 34 s

**Наилучшая модель LogisticRegression**

Наилучшая модель LogisticRegression сохранена в переменной `best_model_clf_3` и имеет следующие 

гиперпераметры:
- max_iter = 20, 

долю правильных ответов:
- accuracy_valid = 0.6967340590979783

была обучена за время:
- 178 ms

Причем,

модель с гиперпараметрами по умолчанию:
- accuracy_valid = 0.6842923794712286
- Wall time: 21 ms

Лучшие показатели accuracy - 0.818 на валидационной выборке показала модель, обученная алгоритмом RandomForestClassifier с подбором трех гиперпараметров затратив на это 21min 15s. Незначительно хуже: accuracy - 0.807, но значительно быстрее - за 1min 59s обучилась модель по алгоритму DecisionTreeClassifier также с подбором трех гиперпараметров. На результат модели, обучаемую алгоритмом LogisticRegression подбор гиперпараметров практически не сказался, для значений по умолчанию accuracy - 0.684, время - 21 ms.

Также отметим, что модель DecisionTreeClassifier уже при подборе 1 гиперпараметра дала приемлемый результат: accuracy - 0.799, время - 183 ms; а модель RandomForestClassifier при значениях по умолчанию: accuracy - 0.785, время - 409 ms.

Исследование моделей проведено, можно приступать к их тестированию.

[В начало раздела 3](#section_3)

<a id="section_4"></a>
## Проверка модели на тестовой выборке

Проверим качество и время работы моделей на тестовой выборке.

In [20]:
%%time
result_test_clf = best_model_clf.score(features_test, target_test)
print(f'Accuracy лучшей модели DecisionTreeClassifier = {result_test_clf:.3f}')

Accuracy лучшей модели DecisionTreeClassifier = 0.782
Wall time: 3.91 ms


In [21]:
%%time
result_test_clf_2 = best_model_clf_2.score(features_test, target_test)
print(f'Accuracy лучшей модели RandomForestClassifier = {result_test_clf_2:.3f}')

Accuracy лучшей модели RandomForestClassifier = 0.785
Wall time: 8.79 ms


In [22]:
%%time
result_test_clf_3 = best_model_clf_3.score(features_test, target_test)
print(f'Accuracy лучшей модели LogisticRegression = {result_test_clf_3:.3f}')

Accuracy лучшей модели LogisticRegression = 0.708
Wall time: 2.93 ms


### Вывод

Результаты проверки работы моделей на тестовой выборке:

Accuracy лучшей модели DecisionTreeClassifier = 0.782, Wall time: 3 ms.

Accuracy лучшей модели RandomForestClassifier = 0.785, Wall time: 10.4 ms.

Accuracy лучшей модели LogisticRegression = 0.708, Wall time: 2.97 ms.

Качество моделей, обученных по алгоритмам DecisionTreeClassifier и RandomForestClassifier удовлетворяет поставленному критерию:  accuracy не ниже 0.75 и практически одинаково, однако время расчета отличается почти в 3.5 раза в пользу DecisionTreeClassifier. Модель, обученая по алгоритму LogisticRegression требуемых показателей не достигла, хотя и показала время расчета идентичное модели DecisionTreeClassifier. 

[В начало раздела 4](#section_4)

<a id="section_5"></a>
## Проверка моделей на вменяемость

Проверим модели на вменяемость.

В разделе "Обзор данных" мы проверяли данные на соответствие здравому смыслу и правдоподобность распределения значений. Тогда мы выяснили, что среднее значение признака `is_ultra` составляет 0.31, т.е. 69% значений равны 0. Таким образом, для проверки моделей на вменяемость составим предсказание из одних нулей и сравним его результат с результатами моделей.

[ссылка](#describe) 

In [23]:
predictions = pd.Series(0, index=target_test.index) 

In [24]:
sanity_check = accuracy_score(target_test, predictions)
print(f'sanity_check = {sanity_check:.3f}')

sanity_check = 0.706


<a id="student_coment_1"></a>
Проверка с использованием DummyClassifier

In [26]:
dummy_clf = DummyClassifier(random_state=12345)
dummy_clf.fit(features_test, target_test)
dummy_clf.score(features_test, target_test)

DummyClassifier(random_state=12345)

0.7060653188180405

### Вывод

Провели проверку моделей на вменяемость.

Рассчитали качество предсказания для случайной выборки - 0.706.

Модель, обученая по алгоритму LogisticRegression, показала результат аналогичный результату случайной выборки, что является неудовлетворительным.

Модели, обученые по алгоритмам DecisionTreeClassifier и RandomForestClassifier, показали результаты, превосходящие результат случайной выборки приблизительно на 8 процентных пунктов, что для нас приемлемо.

[В начало раздела 5](#section_5)

## Итоги исследования

Данное исследование выполнено по запросу оператора мобильной связи «Мегалайн» и будет использовано для построения системы, анализирующей поведение клиентов и рекомендующей, какой из новых тарифов: «Смарт» или «Ультра» - предложить пользователям.
Оператор мобильной связи «Мегалайн» предварительно выяснил, что многие клиенты пользуются архивными тарифами, а компании желательно перевести пользователей на новые.

В нашем распоряжении данные о поведении клиентов, которые уже перешли на новые тарифы: «Смарт» и «Ультра». Предобработка данных уже сделана. Каждый объект в наборе данных — это информация о поведении одного пользователя за месяц. Известно:
- сalls — количество звонков,
- minutes — суммарная длительность звонков в минутах,
- messages — количество sms-сообщений,
- mb_used — израсходованный интернет-трафик в Мб,
- is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

В таблице 3214 объектов по 5 признаков у каждого. Ошибок в данных не выявлено. Количество и качество данных достаточны для исследования.

Исходные данные были нами разделены в соотношении 3:1:1 на обучающую (1928 объектов), валидационную (643 объекта) и тестовую (643 объекта) выборки.

Для построения моделей были использованы 3 алгоритма машинного обучения:
- DecisionTreeClassifier
- RandomForestClassifier
- LogisticRegression

Лучшие показатели accuracy - 0.818 на валидационной выборке показала модель, обученная алгоритмом RandomForestClassifier с подбором трех гиперпараметров и затратив на это 21min 15s. Незначительно хуже: accuracy - 0.807, но значительно быстрее - за 1min 59s обучилась модель по алгоритму DecisionTreeClassifier также с подбором трех гиперпараметров. На результат модели, обучаемой алгоритмом LogisticRegression подбор гиперпараметров практически не сказался, для значений по умолчанию accuracy - 0.684, время - 21 ms.

Также отметим, что модель DecisionTreeClassifier уже при подборе 1 гиперпараметра дала приемлемый результат: accuracy - 0.799, время - 183 ms; а модель RandomForestClassifier при значениях по умолчанию: accuracy - 0.785, время - 409 ms.

Результаты проверки работы моделей на тестовой выборке:

- Accuracy лучшей модели DecisionTreeClassifier = 0.782, Wall time: 3 ms.
- Accuracy лучшей модели RandomForestClassifier = 0.785, Wall time: 10.4 ms.
- Accuracy лучшей модели LogisticRegression = 0.708, Wall time: 2.97 ms.

Качество моделей, обученных по алгоритмам DecisionTreeClassifier и RandomForestClassifier удовлетворяет поставленному критерию:  accuracy не ниже 0.75 и практически одинаково, однако время расчета отличается почти в 3.5 раза в пользу DecisionTreeClassifier. Модель, обученая по алгоритму LogisticRegression требуемых показателей не достигла, хотя и показала время расчета идентичное модели DecisionTreeClassifier. 

Проведена проверка моделей на вменяемость:
- качество предсказания для случайной выборки - 0.706.
- модель, обученая по алгоритму LogisticRegression, показала результат аналогичный результату случайной выборки, что является неудовлетворительным.
- модели, обученые по алгоритмам DecisionTreeClassifier и RandomForestClassifier, показали результаты, превосходящие результат случайной выборки приблизительно на 8 процентных пунктов, что для нас приемлемо.

В результате проведенной работы построены две модели, удовлетворяющих заданному критерию: доля правильных ответов превышает 0.75. Обе модели успешно прошли проверку на вменяемость. В качестве базовой для построения системы, анализирующей поведение клиентов и рекомендующей один из новых тарифов, предпочтительней использовать модель, обученную с использованием алгоритма DecisionTreeClassifier, поскольку при практически одинаковом качестве предсказания её быстродействие гораздо выше.

[В начало](#section_0)

## Чек-лист готовности проекта

На что обратить внимание:

- [x]  Как вы изучаете данные после загрузки?
- [x]  Корректно ли разделяете данные на выборки?
- [x]  Как выбираете размеры выборок?
- [x]  Правильно ли вы оцениваете качество моделей в исследовании?
- [x]  Какие модели и гиперпараметры вы используете?
- [x]  Какие выводы об исследовании делаете?
- [x]  Правильно ли тестируете модели?
- [x]  Насколько высокое значение accuracy получаете?
- [x]  Соблюдаете структуру проекта и поддерживаете аккуратность кода?

[В начало](#section_0)